main/autogen/constant_hash.cc

```diff
@@ -0,0 +1,125 @@
+                hashSoFar = core::mix(hashSoFar, core::_hash(send->method.shortName(gs)));
+                hashSoFar = core::mix(hashSoFar, core::_hash(")"));
+            } else if (send->method == core::Names::include()) {
+                hashSoFar = core::mix(hashSoFar, core::_hash("(i"));
+                hashSoFar = core::mix(hashSoFar, core::_hash(send->method.shortName(gs)));
+                hashSoFar = core::mix(hashSoFar, core::_hash(")"));
+            } else if (send->method == core::Names::extend()) {
+                hashSoFar = core::mix(hashSoFar, core::_hash("(e"));
+                hashSoFar = core::mix(hashSoFar, core::_hash(send->method.shortName(gs)));
+                hashSoFar = core::mix(hashSoFar, core::_hash(")"));
+            }
+        },
+        [&](parser::Assign *asgn) {
+            // we only care about this case if the lhs is a constant
+            if (auto lhs = parser::cast_node<parser::ConstLhs>(asgn->lhs.get())) {
+                if (parser::isa_node<parser::Const>(asgn->rhs.get())) {
+                    // if the RHS is a constant literal, then this is
+                    // an alias and we care about both sides. (This is
+                    // because changing the RHS can, in some perverse
+                    // cases, affect constant resolution: for example,
+                    // in a snippet line
+                    //
+                    //   A = B
+                    //   class C < A; end
+                    //
+                    // which might in turn produce different autogen
+                    // output, i.e. different subclasses lists.)
+                    hashSoFar = core::mix(hashSoFar, core::_hash("(a"));
+                    hashSoFar = core::mix(hashSoFar, core::_hash(lhs->name.shortName(gs)));
+                    hashSoFar = core::mix(hashSoFar, core::_hash(" "));
+                    hashSoFar = hashNode(gs, hashSoFar, asgn->rhs.get());
+                    hashSoFar = core::mix(hashSoFar, core::_hash(")"));
+                } else {
+                    // if the RHS is anything else, then it's a constant
+                    // definition: we care only about the LHS in that case
+                    hashSoFar = core::mix(hashSoFar, core::_hash("(c"));
+                    hashSoFar = core::mix(hashSoFar, core::_hash(lhs->name.shortName(gs)));
+                    hashSoFar = core::mix(hashSoFar, core::_hash(")"));
+                }
+            }
+        },
+        [&](parser::Module *module) {
+            hashSoFar = core::mix(hashSoFar, core::_hash("(m"));
+            hashSoFar = hashNode(gs, hashSoFar, module->name.get());
+            hashSoFar = core::mix(hashSoFar, core::_hash(" "));
+            hashSoFar = hashNode(gs, hashSoFar, module->body.get());
+            hashSoFar = core::mix(hashSoFar, core::_hash(")"));
+        },
+        [&](parser::Class *claz) {
+            hashSoFar = core::mix(hashSoFar, core::_hash("(c"));
```

Comment:
For avoidance of doubt: it doesn't matter that this prefix is also used for LHS constant definitions, right?

Reply:
It should not matter, because classes invariably also add a `<` to their hash (regardless of whether their superclass exists or not) while this constant never will. Might be a good idea to change it just for the mnemonic, though: the specific structure of what we're hashing is mostly for a reader understanding what's going on, because I personally think it's clearer.